### PR TITLE
Feature/orca 309 Build internal_reconcile_report_phantom

### PR DESCRIPTION
**Pull request by AndrewDorn**
## Summary of Changes

Please write a sentence or two summarizing the proposed change.

Addresses [ORCA-309: Build internal_reconcile_report_phantom](https://bugs.earthdata.nasa.gov/browse/ORCA-309)

## Changes

* Added internal_reconcile_report_phantom to code and TF
* Fixed casing in output schema
* Fixed property rename in get_current_archive_list from delete_flag to delete_marker

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## How Has This Been Tested?

Deployed and run in AWS.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the necessary documentation (remove those that do not apply)
    - [x] CHANGELOG.md
    - [x] Testing documentation
    - [x] Development documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
    - [x] Unit tests
- [x] New and existing unit tests pass locally with my changes
    - [x] Automated tests passing (if not fork)
- [x] Any dependent changes have been merged and published in downstream modules
- [x] My code has passed security scanning
    - [x] Snyk
    - [x] git-secrets

### CODE REVIEW

File: modules/lambdas/main.tf [hunk 1]
```diff
@@ -118,6 +118,35 @@ resource "aws_lambda_function" "perform_orca_reconcile" {
...
+  ## REQUIRED
+  function_name = "${var.prefix}_internal_reconcile_report_phantom"
+  role          = var.restore_object_role_arn
+
+  ## OPTIONAL
+  description      = "Returns reconcilliation report sample data."
+  filename         = "${path.module}/../../tasks/internal_reconcile_report_phantom/internal_reconcile_report_phantom.zip"
+  handler          = "internal_reconcile_report_phantom.handler"
+  memory_size      = var.orca_reconciliation_lambda_memory_size
+  runtime          = "python3.7"
```
Comment (rizbih) [inline modules/lambdas/main.tf:133]: I think we want to move over to python 3.9 moving forward

File: tasks/internal_reconcile_report_phantom/internal_reconcile_report_phantom.py [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+
+    return {
+        OUTPUT_JOB_ID_KEY: job_id,
+        OUTPUT_ANOTHER_PAGE_KEY: len(granules) > PAGE_SIZE,
+        OUTPUT_PHANTOMS_KEY: granules[0:PAGE_SIZE],  # we get one extra for anotherPage calculation.
+    }
+
+
+@retry_operational_error()
+def query_db(
```
Comment (rizbih) [inline tasks/internal_reconcile_report_phantom/internal_reconcile_report_phantom.py:55]: should add a minor description of what the function does.

File: tasks/internal_reconcile_report_phantom/internal_reconcile_report_phantom.py [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+                    PHANTOMS_KEY_PATH_KEY: sql_result["key_path"],
+                    PHANTOMS_ORCA_ETAG_KEY: sql_result["orca_etag"],
+                    PHANTOMS_ORCA_LAST_UPDATE_KEY: sql_result["orca_last_update"],
+                    PHANTOMS_ORCA_SIZE_KEY: sql_result["orca_size"],
+                }
+            )
+        return phantoms
+
+
+def get_phantoms_sql() -> text:
```
Comment (rizbih) [inline tasks/internal_reconcile_report_phantom/internal_reconcile_report_phantom.py:96]: should add a minor description of what the function does

File: modules/lambdas/output.tf [hunk 4]
```diff
@@ -26,6 +26,11 @@ output "get_current_archive_list_arn" {
   value       = aws_lambda_function.get_current_archive_list.arn
 }
 
+output "internal_reconcile_report_phantom_arn" {
+  description = "AWS ARN for the internal_reconcile_report_phantom lambda."
+  value       = aws_lambda_function.internal_reconcile_report_phantom.arn
+}
+
```
Comment (rizbih) [inline modules/lambdas/output.tf:33]: should this be added to the modules/main.tf as well similar to https://github.com/nasa/cumulus-orca/blob/4d6cff91690984c4a8b947d89a2115d1e0009301/modules/orca/main.tf#L111 ?

**Review by rizbih — CHANGES_REQUESTED** [review]

File: modules/lambdas/main.tf (hunk 1 above)
Comment (AndrewDorn) [reply modules/lambdas/main.tf:133]: Good catch. Fixed.

File: tasks/internal_reconcile_report_phantom/internal_reconcile_report_phantom.py (hunk 2 above)
Comment (AndrewDorn) [reply tasks/internal_reconcile_report_phantom/internal_reconcile_report_phantom.py:55]: Added

File: tasks/internal_reconcile_report_phantom/internal_reconcile_report_phantom.py (hunk 3 above)
Comment (AndrewDorn) [reply tasks/internal_reconcile_report_phantom/internal_reconcile_report_phantom.py:96]: Added

File: modules/lambdas/output.tf (hunk 4 above)
Comment (AndrewDorn) [reply modules/lambdas/output.tf:33]: I believe that this is not necessary with our current API Gateway architecture.

File: tasks/internal_reconcile_report_phantom/internal_reconcile_report_phantom.py [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+    granule_id, 
+    filename, 
+    key_path, 
+    orca_etag, 
+    (EXTRACT(EPOCH FROM date_trunc('milliseconds', orca_last_update) AT TIME ZONE 'UTC') * 1000)::bigint as orca_last_update,
+    orca_size
+    FROM reconcile_phantom_report
+    WHERE job_id = :job_id
+    ORDER BY collection_id, granule_id, filename
+    OFFSET :page_index*:page_size
```
Comment (bhazuka) [inline tasks/internal_reconcile_report_phantom/internal_reconcile_report_phantom.py:118]: For our initial prototype, this type of paging will work but, it is not performant and as the record size increases this will cause additional slowness. See https://bugs.earthdata.nasa.gov/browse/ORCA-286 as a reference. Ideally, we would want to use a cursor based approach possibly based on collection and granule id. See the following references for the schema to see what a potential schema for cursor and implementation.
- https://relay.dev/graphql/connections.htm#sec-undefined.PageInfo.Fields
- https://graphql.org/learn/pagination/
- https://slack.engineering/evolving-api-pagination-at-slack/
- https://medium.com/swlh/how-to-implement-cursor-pagination-like-a-pro-513140b65f32

Consider creating a performance card to revisit paging like ORCA-286.

File: tasks/internal_reconcile_report_phantom/internal_reconcile_report_phantom.py [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+            [
+                {
+                    "job_id": job_id,
+                    "page_index": page_index,
+                    "page_size": PAGE_SIZE,
+                }
+            ],
+        )
+
+        phantoms = []
```
Comment (bhazuka) [inline tasks/internal_reconcile_report_phantom/internal_reconcile_report_phantom.py:84]: There is nothing necessarily wrong with this approach but we could have the database do all this for us via SQL which may be more performant. By utilizing `array_to_json`, `array_agg`, and `row_to_json` in our PostgreSQL query we could return an array of JSON records without having to step through the records ourselves and put them into a dictionary. For this initial prototype the current implementation is fine. See the following reference for more information on how you could use the functions in SQL.

- https://hashrocket.com/blog/posts/faster-json-generation-with-postgresql

Consider creating a performance card and either add on to the one for the cursor performance mentioned earlier, or create a new one for looking into adjusting the SQL query for performance.

File: tasks/internal_reconcile_report_phantom/test/unit_tests/test_internal_reconcile_report_phantom.py [hunk 7]
```diff
@@ -2,25 +2,284 @@
...
+                    internal_reconcile_report_phantom.PHANTOMS_GRANULE_ID_KEY: granule_id,
+                    internal_reconcile_report_phantom.PHANTOMS_FILENAME_KEY: filename,
+                    internal_reconcile_report_phantom.PHANTOMS_KEY_PATH_KEY: key_path,
+                    internal_reconcile_report_phantom.PHANTOMS_ORCA_ETAG_KEY: orca_etag,
+                    internal_reconcile_report_phantom.PHANTOMS_ORCA_LAST_UPDATE_KEY: orca_last_update,
+                    internal_reconcile_report_phantom.PHANTOMS_ORCA_SIZE_KEY: orca_size,
+                }
+            ],
+            result,
+        )
```
Comment (bhazuka) [inline tasks/internal_reconcile_report_phantom/test/unit_tests/test_internal_reconcile_report_phantom.py:285]: I am wondering if we shouldn't have another "happy" test where in the database returns an empty data set (no phantom records for the given job_id). Thoughts? Does the output schema still hold up if phantoms is an empty array?

File: tasks/internal_reconcile_report_phantom/internal_reconcile_report_phantom.py [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+        for sql_result in sql_results:
```
Comment (bhazuka) [inline tasks/internal_reconcile_report_phantom/internal_reconcile_report_phantom.py:85]: I haven't tested this but any issues if no data returned? In other words no phantoms exist for the given job_id? From a REST perspective, would we return a 404 not found error?

**Review by bhazuka — CHANGES_REQUESTED** [review]
A couple of minor fixes to take care of. I put in some comments for cards for future work unless you feel it should be implemented in this PR. I also posted some questions based on some use cases that should be expected and thoughts to look at. Feel free to let me know if you have questions.

File: modules/lambdas/main.tf [hunk 9]
```diff
@@ -118,6 +118,35 @@ resource "aws_lambda_function" "perform_orca_reconcile" {
...
 
+# internal_reconcile_report_phantom - Receives job id and page index from end user and returns reporting information of files that have records in the ORCA catalog but are missing from S3 bucket.
+# ==============================================================================
+resource "aws_lambda_function" "internal_reconcile_report_phantom" {
+  ## REQUIRED
+  function_name = "${var.prefix}_internal_reconcile_report_phantom"
+  role          = var.restore_object_role_arn
+
+  ## OPTIONAL
+  description      = "Returns reconcilliation report sample data."
```
[the remark this answers is not in the document]
Comment (AndrewDorn) [reply modules/lambdas/main.tf:129]: Whoops. Copied from another lambda. Replaced.

File: modules/lambdas/main.tf [hunk 10]
```diff
@@ -118,6 +118,35 @@ resource "aws_lambda_function" "perform_orca_reconcile" {
...
+  filename         = "${path.module}/../../tasks/internal_reconcile_report_phantom/internal_reconcile_report_phantom.zip"
+  handler          = "internal_reconcile_report_phantom.handler"
+  memory_size      = var.orca_reconciliation_lambda_memory_size
+  runtime          = "python3.9"
```
[the remark this answers is not in the document]
Comment (AndrewDorn) [reply modules/lambdas/main.tf:133]: Ah, I just fixed that in the other branch. Changed.

File: modules/lambdas/main.tf [hunk 11]
```diff
@@ -118,6 +118,35 @@ resource "aws_lambda_function" "perform_orca_reconcile" {
...
+  source_code_hash = filebase64sha256("${path.module}/../../tasks/internal_reconcile_report_phantom/internal_reconcile_report_phantom.zip")
+  tags             = local.tags
```
[the remark this answers is not in the document]
Comment (AndrewDorn) [reply modules/lambdas/main.tf:135]: Fixed

File: tasks/internal_reconcile_report_phantom/internal_reconcile_report_phantom.py (hunk 6 above)
Comment (AndrewDorn) [reply tasks/internal_reconcile_report_phantom/internal_reconcile_report_phantom.py:84]: https://bugs.earthdata.nasa.gov/browse/ORCA-419

File: tasks/internal_reconcile_report_phantom/internal_reconcile_report_phantom.py (hunk 8 above)
Comment (AndrewDorn) [reply tasks/internal_reconcile_report_phantom/internal_reconcile_report_phantom.py:85]: Would seem odd to me to have users wrap each call to our API in try/catch just for the case when no error was found. I'd be open to returning 404 if the job for the job_id is incomplete/non-existent, but that would add an extra sql query.

File: tasks/internal_reconcile_report_phantom/test/unit_tests/test_internal_reconcile_report_phantom.py (hunk 7 above)
Comment (AndrewDorn) [reply tasks/internal_reconcile_report_phantom/test/unit_tests/test_internal_reconcile_report_phantom.py:285]: It is valid. There is no provision for how many items are returned.

File: tasks/internal_reconcile_report_phantom/internal_reconcile_report_phantom.py (hunk 8 above)
Comment (bhazuka) [reply tasks/internal_reconcile_report_phantom/internal_reconcile_report_phantom.py:85]: I don't believe you would need an extra query, you should be able to just check if the result set returned for phantoms is empty.

Comment (AndrewDorn) [reply tasks/internal_reconcile_report_phantom/internal_reconcile_report_phantom.py:85]: Nope. A valid report can have an empty list. Would seem odd to me to have users wrap each call to our API in try/catch just for the case when no error was found.

Comment (bhazuka) [reply tasks/internal_reconcile_report_phantom/internal_reconcile_report_phantom.py:85]: Leave as is for now as we may look at changes down the road if we switch to GraphQL or some other solution.

**Review by rizbih — APPROVED** [review]

**Review by bhazuka — APPROVED** [review]
